internal/server/authn/password_credentials.go

```diff
@@ -0,0 +1,72 @@
+package authn
+
+import (
+	"fmt"
+
+	"golang.org/x/crypto/bcrypt"
+	"gorm.io/gorm"
+
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+)
+
+// passwordCredentialAuthn allows presenting username/password credentials in exchange for an access key
+type passwordCredentialAuthn struct {
+	Username string
+	Password string
+}
+
+func NewPasswordCredentialAuthentication(username, password string) LoginMethod {
+	return &passwordCredentialAuthn{
+		Username: username,
+		Password: password,
+	}
+}
+
+func (a *passwordCredentialAuthn) Authenticate(db *gorm.DB) (*models.Identity, *models.Provider, error) {
+	identity, err := data.GetIdentity(db, data.ByName(a.Username))
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get identity for username: %w", err)
+	}
+
+	// Infra users can have only one username/password combo, look it up
+	userCredential, err := data.GetCredential(db, data.ByIdentityID(identity.ID))
+	if err != nil {
+		return nil, nil, fmt.Errorf("validate creds get user: %w", err)
+	}
+
+	// check if this is a single use password that was already used
+	if userCredential.OneTimePassword && userCredential.OneTimePasswordUsed {
+		return nil, nil, fmt.Errorf("one time password cannot be used more than once")
```

Comment:
add some guidance? "have an administrator reset your password" ?

Reply:
This error doesn't reach the user in the client, so more details dont really help here. We can revisit this in a separate change.